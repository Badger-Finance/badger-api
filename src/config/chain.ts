import { ethers } from 'ethers';
import { SettDefinition } from '../interface/Sett';
import { Token } from '../interface/Token';
import { Provider } from './constants';
import { ethSetts } from './setts/eth-setts';

export interface Chain {
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly tokens: Token[];
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
  public name = 'Ethereum';
  public symbol = 'eth';
  public chainId = '0x01';
  public tokens = [];
  public setts = ethSetts;
  public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
  public name = 'BinanceSmartChain';
  public symbol = 'bsc';
  public chainId = '0x61';
  public tokens = [];
  public setts = [];
  public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 * TODO: Should not supported throw a bad request instead?
 */
export const resolveChainQuery = (symbol: string): Chain => {
  if (!symbol) return eth;
  const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
  if (!chain) return eth;
  return chain;
};
