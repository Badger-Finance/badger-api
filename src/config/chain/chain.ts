import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { SettDefinition } from '../../interface/Sett';
import { Token } from '../../interface/Token';
import { Provider } from '../constants';
import { bscSetts } from './bsc';
import { ethSetts } from './eth';

export interface Chain {
	readonly name: string;
	readonly symbol: string;
	readonly chainId: string;
	readonly tokens: Token[];
	readonly setts: SettDefinition[];
	readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
	public name = 'Ethereum';
	public symbol = 'eth';
	public chainId = '0x01';
	public tokens = [];
	public setts = ethSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
	public name = 'BinanceSmartChain';
	public symbol = 'bsc';
	public chainId = '0x61';
	public tokens = [];
	public setts = bscSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 */
export const resolveChainQuery = (symbol: string): Chain => {
	if (!symbol) return eth;
	const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
	if (!chain) throw new BadRequest('Invalid chain');
	return chain;
};
