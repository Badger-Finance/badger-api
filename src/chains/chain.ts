import { Arbitrum } from './config/arbitrum.config';
import { BinanceSmartChain } from './config/bsc.config';
import { Chain } from './config/chain.config';
import { Ethereum } from './config/eth.config';
import { Polygon } from './config/matic.config';

/**
 * Instantiate objects for registration.
 */
export const loadChains = (): Chain[] => {
  return [new Ethereum(), new BinanceSmartChain(), new Polygon(), new Arbitrum()];
};
