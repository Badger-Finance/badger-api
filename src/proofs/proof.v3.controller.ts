import { Network } from '@badger-dao/sdk';
import { MerkleProof } from '@badger-dao/sdk/lib/api/types/merkle-proof';
import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { QueryParamError } from '../errors/validation/query.param.error';
import { ProofsService } from './proofs.service';

@Controller('/proof')
export class ProofsV3Controller {
  @Inject()
  proofsService!: ProofsService;

  @Get()
  @ContentType('json')
  async getBouncerProof(
    @QueryParams('address') address: string,
    @QueryParams('chain') chain?: Network,
  ): Promise<MerkleProof> {
    if (!address) throw new QueryParamError('address');

    return this.proofsService.getBouncerProof(Chain.getChain(chain), address);
  }
}
