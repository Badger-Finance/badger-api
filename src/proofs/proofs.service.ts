import { MerkleProof } from '@badger-dao/sdk';
import { Service } from '@tsed/common';
import { ethers } from 'ethers';

import { getObject } from '../aws/s3.utils';
import { Chain } from '../chains/config/chain.config';
import { REWARD_DATA } from '../config/constants';
import { NodataForAddrError } from '../errors/allocation/nodata.for.addr.error';
import { NodataForChainError } from '../errors/allocation/nodata.for.chain.error';
import { AirdropMerkleDistribution } from '../rewards/interfaces/merkle-distributor.interface';

@Service()
export class ProofsService {
  /**
   * Load user bouncer proof. These proofs are used for vault guest lists.
   * @param address User Ethereum address.
   */
  async getBouncerProof(chain: Chain, address: string): Promise<MerkleProof> {
    const fileName = `badger-bouncer-${parseInt(chain.chainId)}.json`;
    const bouncerFile = await getObject(REWARD_DATA, fileName);
    if (!bouncerFile) {
      throw new NodataForChainError(chain.name);
    }
    const fileContents: AirdropMerkleDistribution = JSON.parse(bouncerFile.toString('utf-8'));
    const claim = fileContents.claims[ethers.utils.getAddress(address)];
    if (!claim) {
      throw new NodataForAddrError(`${address}`);
    }
    return claim.proof;
  }
}
