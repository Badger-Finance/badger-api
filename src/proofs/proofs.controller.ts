import { Network } from '@badger-dao/sdk';
import { MerkleProof } from '@badger-dao/sdk/lib/api/types/merkle-proof';
import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { ProofsService } from './proofs.service';
import { CitadelMerkleClaim } from './interfaces/citadel-merkle-claim.interface';

@Controller('/proofs')
export class ProofsController {
  @Inject()
  proofsService!: ProofsService;

  @Get('/:address')
  @ContentType('json')
  async getBouncerProof(
    @PathParams('address') address: string,
    @QueryParams('chain') chain?: Network,
  ): Promise<MerkleProof> {
    return this.proofsService.getBouncerProof(Chain.getChain(chain), address);
  }

  @Get('/citadel/:address')
  @ContentType('json')
  async getCitadelProof(@PathParams('address') address: string): Promise<CitadelMerkleClaim> {
    return this.proofsService.getCitadelClaim(address);
  }
}
