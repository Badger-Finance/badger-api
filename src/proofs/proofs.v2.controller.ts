import { Network } from '@badger-dao/sdk';
import { MerkleProof } from '@badger-dao/sdk/lib/api/types/merkle-proof';
import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType, Deprecated } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { ProofsService } from './proofs.service';

@Deprecated()
@Controller('/proofs')
export class ProofsV2Controller {
  @Inject()
  proofsService!: ProofsService;

  @Get('/:address')
  @ContentType('json')
  async getBouncerProof(
    @PathParams('address') address: string,
    @QueryParams('chain') chain?: Network,
  ): Promise<MerkleProof> {
    return this.proofsService.getBouncerProof(Chain.getChain(chain), address);
  }
}
