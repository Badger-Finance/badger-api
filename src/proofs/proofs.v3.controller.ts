import { MerkleProof, Network } from '@badger-dao/sdk';
import { Controller, Inject } from '@tsed/di';
import { QueryParams } from '@tsed/platform-params';
import { ContentType, Get } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { QueryParamError } from '../errors/validation/query.param.error';
import { ProofsService } from './proofs.service';

@Controller('/proofs')
export class ProofsV3Controller {
  @Inject()
  proofsService!: ProofsService;

  @Get()
  @ContentType('json')
  async getBouncerProof(
    @QueryParams('address') address: string,
    @QueryParams('chain') chain?: Network
  ): Promise<MerkleProof> {
    if (!address) {
      throw new QueryParamError('address');
    }
    return this.proofsService.getBouncerProof(Chain.getChain(chain), address);
  }
}
