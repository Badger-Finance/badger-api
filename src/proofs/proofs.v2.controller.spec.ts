import { NotFound } from '@tsed/exceptions';
import { PlatformServerless } from '@tsed/platform-serverless';
import { PlatformServerlessTest } from '@tsed/platform-serverless-testing';

import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import { NodataForChainError } from '../errors/allocation/nodata.for.chain.error';
import { NetworkStatus } from '../errors/enums/network-status.enum';
import { MOCK_BOUNCER_FILE, TEST_ADDR } from '../test/constants';
import { setupMockChain } from '../test/mocks.utils';
import { ProofsService } from './proofs.service';
import { ProofsV2Controller } from './proofs.v2.controller';

describe('ProofsController', () => {
  beforeEach(
    PlatformServerlessTest.bootstrap(PlatformServerless, {
      lambda: [ProofsV2Controller],
    }),
  );
  afterEach(() => PlatformServerlessTest.reset());

  beforeEach(setupMockChain);

  describe('GET /v2/proofs', () => {
    it('returns 404 for a chain with no bouncer file', async () => {
      jest.spyOn(ProofsService.prototype, 'getBouncerProof').mockImplementation(async (chain) => {
        // simulate no chain path
        throw new NodataForChainError(chain.network);
      });
      const { body, statusCode } = await PlatformServerlessTest.request.get(`/proofs/${TEST_ADDR}`);
      expect(statusCode).toEqual(NetworkStatus.NotFound);
      expect(JSON.parse(body)).toMatchSnapshot();
    });

    it('returns 404 for users not on the bouncer list', async () => {
      const badAddress = TOKENS.BADGER;
      jest.spyOn(ProofsService.prototype, 'getBouncerProof').mockImplementation(async () => {
        // simulate no user proofs path
        throw new NotFound(`${badAddress} is not on the bouncer list`);
      });
      const { body, statusCode } = await PlatformServerlessTest.request.get(`/proofs/${badAddress}`);
      expect(statusCode).toEqual(NetworkStatus.NotFound);
      expect(JSON.parse(body)).toMatchSnapshot();
    });

    it('returns 200 and the merkle proof for a user on the bouncer list', async () => {
      jest
        .spyOn(ProofsService.prototype, 'getBouncerProof')
        .mockImplementation(async (_chain: Chain, _address: string) => MOCK_BOUNCER_FILE.claims[TEST_ADDR].proof);
      const { body, statusCode } = await PlatformServerlessTest.request.get(`/proofs/${TEST_ADDR}`);
      expect(statusCode).toEqual(NetworkStatus.Success);
      expect(JSON.parse(body)).toMatchSnapshot();
    });
  });
});
