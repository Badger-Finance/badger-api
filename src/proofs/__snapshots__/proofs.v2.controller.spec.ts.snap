// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProofsController GET /v2/proofs returns 200 and the merkle proof for a user on the bouncer list 1`] = `
Array [
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
]
`;

exports[`ProofsController GET /v2/proofs returns 404 for a chain with no bouncer file 1`] = `
Object {
  "message": "Ethereum does not have a bouncer list",
}
`;

exports[`ProofsController GET /v2/proofs returns 404 for users not on the bouncer list 1`] = `
Object {
  "message": "0x3472A5A71965499acd81997a54BBA8D852C6E53d is not on the bouncer list",
}
`;
